desc/magnetic_fields.py

```diff
@@ -1846,6 +1853,328 @@ def from_surface(
+        array of shape (L_B + 1, M_B), where the rows are Chebyshev coefficients
+        corresponding to the modes in `B_basis` for the radial variation, and the
+        columns are the values of ||B|| atlinearly spaced monotonic spline knots.
+        If not supplied, `B_lm` defaults to a constant field of 1 T.
+    x_lmn : ndarray, optional
+        Omnigenity parameters describing h(ρ,η,α). The coefficients correspond to the
+        modes in `x_basis`. If not supplied, `x_lmn` defaults to zero for all modes.
+
+    """
+
+    _io_attrs_ = [
+        "_L_B",
+        "_M_B",
+        "_L_x",
+        "_M_x",
+        "_N_x",
+        "_NFP",
+        "_helicity",
+        "_B_basis",
+        "_x_basis",
+        "_B_lm",
+        "_x_lmn",
+    ]
+
+    def __init__(
+        self,
+        L_B=0,
+        M_B=2,
+        L_x=0,
+        M_x=0,
+        N_x=0,
+        NFP=1,
+        helicity=(1, 0),
+        B_lm=None,
+        x_lmn=None,
+    ):
+        self._L_B = int(L_B)
+        self._M_B = int(M_B)
+        self._L_x = int(L_x)
+        self._M_x = int(M_x)
+        self._N_x = int(N_x)
+        self._NFP = int(NFP)
+        self.helicity = helicity
+        self._B_basis = ChebyshevPolynomial(L=self.L_B)
+        self._x_basis = ChebyshevDoubleFourierBasis(
+            L=self.L_x,
+            M=self.M_x,
+            N=self.N_x,
+            NFP=self.NFP,
+            sym="cos(t)",
+        )
+        if B_lm is None:
+            self._B_lm = np.concatenate(
+                (
+                    np.ones((self.M_B,)),  # constant |B| = 1 T
+                    np.zeros((self.L_B * self.M_B,)),  # same field on all flux surfaces
+                )
+            )
+        else:
+            assert len(B_lm) == (self.L_B + 1) * self.M_B
+            self._B_lm = B_lm
+        if x_lmn is None:
+            self._x_lmn = np.zeros(self.x_basis.num_modes)
+        else:
+            assert len(x_lmn) == self.x_basis.num_modes
+            self._x_lmn = x_lmn
+
+        # TODO: should we not allow some types of helicity?
+        helicity_sign = sign(helicity[0]) * sign(helicity[1])
+        warnif(
+            self.helicity != (0, self.NFP * helicity_sign)
+            and abs(self.helicity[0]) != 1,
+            UserWarning,
+            "Typical helicity (M,N) has M=1.",
+        )
+        warnif(
+            self.helicity != (helicity_sign, 0) and abs(self.helicity[1]) != self.NFP,
+            UserWarning,
+            "Typical helicity (M,N) has N=NFP.",
+        )
+
+    def change_resolution(
+        self,
+        L_B=None,
+        M_B=None,
+        L_x=None,
+        M_x=None,
+        N_x=None,
+        NFP=None,
+    ):
+        """Set the spectral resolution of well and shift terms.
+
+        Parameters
+        ----------
+        L_B : int
+            Radial resolution of the magnetic well parameters B_lm.
+        M_B : int
+            Number of spline points in the magnetic well parameters B_lm.
```

Comment:
make these consistent with your definitions of these paremeters in the `init` i.e. there you say `M_B` is number of spline knots per surface, so might as well keep that level of description in the change resolution as well